### DISCUSSION TITLE
Does uvicorn support pathsend?

### DISCUSSION THREAD

**Discussion by multimeric** [Kludex/uvicorn#2286] [Q&A] [answered]
It seems that starlette now supports pathsend: https://github.com/encode/starlette/pull/2435. However, using this feature requires an ASGI server that also supports pathsend. Does uvicorn?

**Comment by gi0baro** [comment, marked as answer]
@multimeric `pathsend` was proposed and added to ASGI spec by me to allow ASGI frameworks/apps to benefit from servers implementing such standard. The proposal was mainly driven by the existence of a similar behaviour in [RSGI protocol](https://github.com/emmett-framework/granian/blob/master/docs/spec/RSGI.md).

Uvicorn doesn't support it, but it probably won't benefit at all from it, as `pathsend` is mainly designed for servers which are not written in Python. In order to support `pathsend` Uvicorn would implement the same Python code your application/framework already implement, behaving in the exact same way, and being limited by the same stack (GIL, Python objects, etc.).

**Reply by multimeric** [reply]
Thanks, so the benefits of `pathsend` will mostly be when using an ASGI server that can handle static files faster than Python?

**Reply by gi0baro** [reply]
I think the point is more about concurrency rather than speed. A pure Python `pathsend` implementation would still hold the GIL in the same way the application/framework does.

**Reply by Archmonger** [reply]
I'm not sure if I agree with this. The efficiency brought from `pathsend` is dependent on implementation. Pure-python ASGI webservers could implement pathsend as something akin to `asyncio.to_thread( os.sendfile, ... )`. When wrapping the operating system's syscall in a thread, it would absolutely would free up a ton of GIL time (_especially_ for huge file transfers) compared to reading/sending each chunk in Python. In theory, this could open the doors to new ASGI use cases (e.g. video streaming).

If `os.sendfile` doesn't exist ([such as on Windows](https://github.com/python/cpython/pull/112337)) then it would need to fall-back to traditional python chunking.

@Kludex I was thinking of PRing this if you agree.

**Reply by gi0baro** [reply]
You can't use `os.sendfile`, the socket is handled by the event loop.

**Reply by Archmonger** [reply]
That was more of a pseudo-code example. I just drafted a functioning prototype with `loop.sendfile` (which uses `os.sendfile` under the hood). The performance implements ultimately relies on the user's event loop having built-in concurrency methodology for the syscall, which I believe is an acceptable expectation.

See PR https://github.com/Kludex/uvicorn/pull/2840 for implementation.
